cpp/src/arrow/compute/kernels/vector_hash.cc

```diff
@@ -451,8 +487,12 @@ struct HashKernelTraits<Type, Action, enable_if_has_string_view<Type>> {
 template <typename Type, typename Action>
 std::unique_ptr<HashKernel> HashInitImpl(KernelContext* ctx, const KernelInitArgs& args) {
   using HashKernelType = typename HashKernelTraits<Type, Action>::HashKernel;
-  auto result = ::arrow::internal::make_unique<HashKernelType>(args.inputs[0].type,
-                                                               ctx->memory_pool());
+  DictionaryEncodeOptions options;
+  if (auto options_ptr = static_cast<const DictionaryEncodeOptions*>(args.options)) {
```

Comment:
This seems fragile; if other functions which use `HashInitImpl` add options this will break. Would it be possible to branch on the Action instead?

```suggestion
  if (std::is_same<Action, DictEncodeAction::value) {
```

(currently dictionary_encode uses ValueCountsAction in Init, which is confusing and would invalidate the above approach)

Reply:
I moved the options decoding into Action although the `HashKernel` still needed to be aware of it (to decide if `null` should go in the dictionary or not) so it changed the kernel->action interface slightly (added `ShouldEncodeNulls`)